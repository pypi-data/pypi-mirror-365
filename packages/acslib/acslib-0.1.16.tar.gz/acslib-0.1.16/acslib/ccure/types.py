from enum import Enum


class ObjectType(Enum):
    CLEARANCE = "clearance"
    CLEARANCE_ASSIGNMENT = "clearance assignment"
    CREDENTIAL = "credential"
    CLEARANCE_ITEM = "clearance item"
    DOOR = "door"
    ISTAR_DOOR = "istar door"
    ELEVATOR = "elevator"
    IMAGE = "image"
    ISTAR_CONTROLLER = "istar controller"
    PERSONNEL = "personnel"
    TIME_SPEC = "time spec"
    GROUP = "group"
    GROUP_MEMBER = "group member"

    @property
    def complete(self):
        if self == self.CLEARANCE:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Clearance"
        if self == self.CLEARANCE_ASSIGNMENT:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.PersonnelClearancePair"
        if self == self.CREDENTIAL:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Credential"
        if self == self.CLEARANCE_ITEM:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.ClearanceItem"
        if self == self.DOOR:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Door"
        if self == self.ISTAR_DOOR:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.iStarDoor"
        if self == self.ELEVATOR:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Elevator"
        if self == self.IMAGE:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Images"
        if self == self.ISTAR_CONTROLLER:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.iStarController"
        if self == self.PERSONNEL:
            return "SoftwareHouse.NextGen.Common.SecurityObjects.Personnel"
        if self == self.TIME_SPEC:
            return "SoftwareHouse.CrossFire.Common.Objects.TimeSpec"
        if self == self.GROUP:
            return "SoftwareHouse.CrossFire.Common.Objects.Group"
        if self == self.GROUP_MEMBER:
            return "SoftwareHouse.CrossFire.Common.Objects.GroupMember"


class ImageType(Enum):
    UNKNOWN = 0
    PORTRAIT = 1
    SIGNATURE = 2
    FINGERPRINT = 3
    HANDPRINT = 4
    DYNAMIC_BADGE_IMAGE = 5
    STATIC_BADGE_IMAGE = 6
    SYSTEM_IMAGE = 7
    PRIVATE_IMAGE = 8
    SHARED_IMAGE = 9
