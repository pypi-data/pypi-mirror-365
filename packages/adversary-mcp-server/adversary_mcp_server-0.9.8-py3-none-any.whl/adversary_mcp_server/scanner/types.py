"""Core data types for security vulnerability detection."""

import uuid
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path


class Severity(str, Enum):
    """Security vulnerability severity levels."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class Category(str, Enum):
    """Security vulnerability categories."""

    INJECTION = "injection"
    AUTHENTICATION = "authentication"
    AUTHORIZATION = "authorization"
    CRYPTO = "crypto"
    CRYPTOGRAPHY = "cryptography"
    CONFIGURATION = "configuration"
    VALIDATION = "validation"
    LOGGING = "logging"
    DESERIALIZATION = "deserialization"
    SSRF = "ssrf"
    XSS = "xss"
    IDOR = "idor"
    RCE = "rce"
    LFI = "lfi"
    DISCLOSURE = "disclosure"
    ACCESS_CONTROL = "access_control"
    TYPE_SAFETY = "type_safety"
    SECRETS = "secrets"
    DOS = "dos"
    CSRF = "csrf"
    PATH_TRAVERSAL = "path_traversal"
    REDIRECT = "redirect"
    HEADERS = "headers"
    SESSION = "session"
    FILE_UPLOAD = "file_upload"
    XXE = "xxe"
    CLICKJACKING = "clickjacking"
    MISC = "misc"  # Generic category for miscellaneous threats


class Language(str, Enum):
    """Supported programming languages."""

    PYTHON = "python"
    JAVASCRIPT = "javascript"
    TYPESCRIPT = "typescript"
    HTML = "html"
    CSS = "css"
    JSON = "json"
    YAML = "yaml"
    XML = "xml"
    PHP = "php"
    SHELL = "shell"
    DOCKERFILE = "dockerfile"
    GO = "go"
    RUBY = "ruby"
    JAVA = "java"
    CSHARP = "csharp"
    SQL = "sql"
    TERRAFORM = "terraform"
    GENERIC = "generic"  # For unknown file types


class LanguageSupport:
    """Centralized language support configuration.

    This class contains all language-related mappings and configurations
    in one place to avoid duplication across modules.
    """

    # Central configuration: Language -> (primary_extension, [all_extensions])
    _LANGUAGE_CONFIG = {
        Language.PYTHON: (".py", [".py", ".pyw", ".pyx"]),
        Language.JAVASCRIPT: (".js", [".js", ".jsx", ".mjs", ".cjs"]),
        Language.TYPESCRIPT: (".ts", [".ts", ".tsx", ".mts", ".cts"]),
        Language.HTML: (
            ".html",
            [
                ".html",
                ".htm",
                ".ejs",
                ".handlebars",
                ".hbs",
                ".mustache",
                ".vue",
                ".svelte",
            ],
        ),
        Language.CSS: (".css", [".css", ".scss", ".sass", ".less", ".stylus"]),
        Language.JSON: (".json", [".json", ".jsonc", ".json5"]),
        Language.YAML: (".yaml", [".yaml", ".yml"]),
        Language.XML: (".xml", [".xml", ".xsl", ".xslt", ".svg", ".rss", ".atom"]),
        Language.PHP: (
            ".php",
            [".php", ".phtml", ".php3", ".php4", ".php5", ".php7", ".phps"],
        ),
        Language.SHELL: (".sh", [".sh", ".bash", ".zsh", ".fish", ".ksh", ".csh"]),
        Language.DOCKERFILE: (".dockerfile", [".dockerfile", "Dockerfile"]),
        Language.GO: (".go", [".go"]),
        Language.RUBY: (".rb", [".rb", ".rbw", ".rake", ".gemspec"]),
        Language.JAVA: (".java", [".java", ".groovy", ".gradle"]),
        Language.CSHARP: (".cs", [".cs", ".csx", ".vb"]),
        Language.SQL: (".sql", [".sql", ".psql", ".mysql", ".sqlite", ".plsql"]),
        Language.TERRAFORM: (".tf", [".tf", ".tfvars", ".hcl"]),
        Language.GENERIC: (
            ".txt",
            [
                ".txt",
                ".md",
                ".cfg",
                ".conf",
                ".ini",
                ".env",
                ".toml",
                ".log",
                ".properties",
            ],
        ),
    }

    @classmethod
    def get_supported_languages(cls) -> list[Language]:
        """Get list of all supported languages."""
        return list(cls._LANGUAGE_CONFIG.keys())

    @classmethod
    def get_extension_to_language_map(cls) -> dict[str, Language]:
        """Get mapping from file extensions to languages."""
        extension_map = {}
        for language, (_, extensions) in cls._LANGUAGE_CONFIG.items():
            for ext in extensions:
                # Handle special case for Dockerfile without extension
                if ext == "Dockerfile":
                    extension_map["dockerfile"] = language  # lowercase for matching
                else:
                    extension_map[ext] = language
        return extension_map

    @classmethod
    def get_language_to_extension_map(cls) -> dict[Language, str]:
        """Get mapping from languages to their primary file extensions."""
        return {
            language: primary_ext
            for language, (primary_ext, _) in cls._LANGUAGE_CONFIG.items()
        }

    @classmethod
    def detect_language(cls, file_path: str | Path) -> Language:
        """Detect language from file path."""
        path = Path(file_path)

        # Handle special case for Dockerfile
        if path.name.lower() in ["dockerfile", "containerfile"]:
            return Language.DOCKERFILE

        extension = path.suffix.lower()
        extension_map = cls.get_extension_to_language_map()

        return extension_map.get(extension, Language.GENERIC)

    @classmethod
    def get_extensions_for_language(cls, language: Language) -> list[str]:
        """Get all file extensions for a specific language."""
        if language in cls._LANGUAGE_CONFIG:
            return cls._LANGUAGE_CONFIG[language][1].copy()
        return []

    @classmethod
    def get_primary_extension(cls, language: Language) -> str:
        """Get the primary file extension for a language."""
        if language in cls._LANGUAGE_CONFIG:
            return cls._LANGUAGE_CONFIG[language][0]
        return ".txt"

    @classmethod
    def get_language_enum_values(cls) -> list[str]:
        """Get list of language enum values for API schemas."""
        return [lang.value for lang in cls.get_supported_languages()]


@dataclass
class ThreatMatch:
    """A detected security threat."""

    # Required fields
    rule_id: str
    rule_name: str
    description: str
    category: Category
    severity: Severity
    file_path: str
    line_number: int

    # Optional fields with defaults
    uuid: str = field(default_factory=lambda: str(uuid.uuid4()))
    column_number: int = 0
    code_snippet: str = ""
    function_name: str | None = None
    exploit_examples: list[str] = field(default_factory=list)
    remediation: str = ""
    references: list[str] = field(default_factory=list)
    cwe_id: str | None = None
    owasp_category: str | None = None
    confidence: float = 1.0  # 0.0 to 1.0
    source: str = "rules"  # Scanner source: "rules", "semgrep", "llm"
    is_false_positive: bool = False  # False positive tracking

    def get_fingerprint(self) -> str:
        """Generate a unique fingerprint for this finding.

        Used to identify the same logical finding across multiple scans
        to preserve UUIDs and false positive markings.

        Returns:
            Unique fingerprint string based on rule_id, file_path, and line_number
        """
        from pathlib import Path

        # Normalize file path to handle relative vs absolute paths
        normalized_path = str(Path(self.file_path).resolve())
        return f"{self.rule_id}:{normalized_path}:{self.line_number}"
