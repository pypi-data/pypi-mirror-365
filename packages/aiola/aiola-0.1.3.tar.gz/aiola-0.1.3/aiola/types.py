from __future__ import annotations

import enum
from collections.abc import Mapping
from dataclasses import dataclass
from typing import IO, TypedDict, Union

from .constants import DEFAULT_AUTH_BASE_URL, DEFAULT_BASE_URL, DEFAULT_HTTP_TIMEOUT, DEFAULT_WORKFLOW_ID


@dataclass
class AiolaClientOptions:
    """Configuration options for Aiola clients."""

    base_url: str | None = DEFAULT_BASE_URL
    auth_base_url: str | None = DEFAULT_AUTH_BASE_URL
    api_key: str | None = None
    access_token: str | None = None
    workflow_id: str = DEFAULT_WORKFLOW_ID
    timeout: float | None = DEFAULT_HTTP_TIMEOUT

    def __post_init__(self) -> None:
        """Validate options after initialization."""
        if not self.api_key and not self.access_token:
            raise ValueError("Either api_key or access_token must be provided")

        if self.api_key is not None and not isinstance(self.api_key, str):
            raise TypeError("API key must be a string")

        if self.access_token is not None and not isinstance(self.access_token, str):
            raise TypeError("Access token must be a string")

        if self.base_url is not None and not isinstance(self.base_url, str):
            raise TypeError("Base URL must be a string")

        if self.auth_base_url is not None and not isinstance(self.auth_base_url, str):
            raise TypeError("Auth base URL must be a string")

        if not isinstance(self.workflow_id, str):
            raise TypeError("Workflow ID must be a string")

        if self.timeout is not None and not isinstance(self.timeout, (int | float)):
            raise TypeError("Timeout must be a number")


class LiveEvents(str, enum.Enum):
    Transcript = "transcript"
    Translation = "translation"
    SentimentAnalysis = "sentiment_analysis"
    Summarization = "summarization"
    TopicDetection = "topic_detection"
    ContentModeration = "content_moderation"
    AutoChapters = "auto_chapters"
    FormFilling = "form_filling"
    EntityDetection = "entity_detection"
    EntityDetectionFromList = "entity_detection_from_list"
    KeyPhrases = "key_phrases"
    PiiRedaction = "pii_redaction"
    Error = "error"
    Disconnect = "disconnect"
    Connect = "connect"


class Segment(TypedDict):
    start: float
    end: float
    text: str


class TranscriptionMetadata(TypedDict):
    """Metadata for transcription results."""

    duration: float
    language: str
    sample_rate: int
    num_channels: int
    timestamp_utc: str
    model_version: str


class TranscriptionResponse(TypedDict):
    """Response from file transcription API."""

    transcript: str
    itn_transcript: str
    segments: list[Segment]
    metadata: TranscriptionMetadata


class SessionCloseResponse(TypedDict):
    """Response from session close API."""

    status: str
    deletedAt: str


class GrantTokenResponse(TypedDict):
    """Response from grant token API."""

    accessToken: str
    sessionId: str


class TranslationPayload(TypedDict):
    src_lang_code: str
    dst_lang_code: str


class EntityDetectionFromListPayload(TypedDict):
    entity_list: list[str]


class _EmptyPayload(TypedDict):
    pass


EntityDetectionPayload = _EmptyPayload
KeyPhrasesPayload = _EmptyPayload
PiiRedactionPayload = _EmptyPayload
SentimentAnalysisPayload = _EmptyPayload
SummarizationPayload = _EmptyPayload
TopicDetectionPayload = _EmptyPayload
ContentModerationPayload = _EmptyPayload
AutoChaptersPayload = _EmptyPayload
FormFillingPayload = _EmptyPayload


class TasksConfig(TypedDict, total=False):
    FORM_FILLING: FormFillingPayload
    TRANSLATION: TranslationPayload
    ENTITY_DETECTION: EntityDetectionPayload
    ENTITY_DETECTION_FROM_LIST: EntityDetectionFromListPayload
    KEY_PHRASES: KeyPhrasesPayload
    PII_REDACTION: PiiRedactionPayload
    SENTIMENT_ANALYSIS: SentimentAnalysisPayload
    SUMMARIZATION: SummarizationPayload
    TOPIC_DETECTION: TopicDetectionPayload
    CONTENT_MODERATION: ContentModerationPayload
    AUTO_CHAPTERS: AutoChaptersPayload


FileContent = Union[IO[bytes], bytes, str]
File = Union[
    # file (or bytes)
    FileContent,
    # (filename, file (or bytes))
    tuple[str | None, FileContent],
    # (filename, file (or bytes), content_type)
    tuple[str | None, FileContent, str | None],
    # (filename, file (or bytes), content_type, headers)
    tuple[
        str | None,
        FileContent,
        str | None,
        Mapping[str, str],
    ],
]
