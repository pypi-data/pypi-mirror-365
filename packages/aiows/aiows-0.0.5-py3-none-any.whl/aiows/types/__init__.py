"""
Type definitions for aiows
"""

from datetime import datetime
from typing import Literal, Tuple
from pydantic import BaseModel, Field


class BaseMessage(BaseModel):
    """Base class for all WebSocket messages"""
    type: str
    timestamp: datetime = Field(default_factory=datetime.now)
    
    def dict(self, **kwargs):
        """Serialize message to dictionary"""
        return super().model_dump(**kwargs)


class ChatMessage(BaseMessage):
    """Chat message with text content"""
    type: Literal["chat"] = "chat"
    text: str
    user_id: int


class JoinRoomMessage(BaseMessage):
    """Message for joining a room"""
    type: Literal["join_room"] = "join_room"
    room_id: str
    user_name: str


class GameActionMessage(BaseMessage):
    """Game action message with coordinates"""
    type: Literal["game_action"] = "game_action"
    action: str
    coordinates: Tuple[int, int] 