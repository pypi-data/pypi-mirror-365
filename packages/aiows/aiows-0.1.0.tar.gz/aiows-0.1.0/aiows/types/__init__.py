"""
Type definitions for aiows
"""

from datetime import datetime
from typing import Literal, Tuple
from pydantic import BaseModel, Field, field_validator

from ..validators import (
    validate_safe_text, validate_username, validate_room_id, 
    validate_game_action, validate_user_id, validate_coordinates
)


class BaseMessage(BaseModel):
    """Base class for all WebSocket messages"""
    type: str
    timestamp: datetime = Field(default_factory=datetime.now)
    
    def dict(self, **kwargs):
        """Serialize message to dictionary"""
        return super().model_dump(**kwargs)


class ChatMessage(BaseMessage):
    """Chat message with text content"""
    type: Literal["chat"] = "chat"
    text: str
    user_id: int
    
    @field_validator('text')
    @classmethod
    def validate_text_field(cls, v):
        return validate_safe_text(cls, v)
    
    @field_validator('user_id', mode='before')
    @classmethod
    def validate_user_id_field(cls, v):
        return validate_user_id(cls, v)


class JoinRoomMessage(BaseMessage):
    """Message for joining a room"""
    type: Literal["join_room"] = "join_room"
    room_id: str
    user_name: str
    
    @field_validator('room_id')
    @classmethod
    def validate_room_id_field(cls, v):
        return validate_room_id(cls, v)
    
    @field_validator('user_name')
    @classmethod
    def validate_user_name_field(cls, v):
        return validate_username(cls, v)


class GameActionMessage(BaseMessage):
    """Game action message with coordinates"""
    type: Literal["game_action"] = "game_action"
    action: str
    coordinates: Tuple[int, int]
    
    @field_validator('action')
    @classmethod
    def validate_action_field(cls, v):
        return validate_game_action(cls, v)
    
    @field_validator('coordinates')
    @classmethod
    def validate_coordinates_field(cls, v):
        return validate_coordinates(cls, v)