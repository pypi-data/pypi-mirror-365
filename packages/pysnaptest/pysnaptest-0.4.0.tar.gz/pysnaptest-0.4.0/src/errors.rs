use std::env::VarError;
use std::fmt::{self, Display, Formatter};

use pyo3::exceptions::PyValueError;
use pyo3::PyErr;
use pythonize::PythonizeError;

#[derive(Debug)]
pub enum PytestInfoError {
    CouldNotSplit(String),
    InvalidEnvVar(VarError),
    NoTestFile,
}

impl Display for PytestInfoError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            PytestInfoError::CouldNotSplit(s) => {
                write!(f, "Expected '::' to be in PYTEST_CURRENT_TEST string ({s})")
            }
            PytestInfoError::InvalidEnvVar(e) => match e {
                VarError::NotPresent => write!(f, "PYTEST_CURRENT_TEST is not set"),
                VarError::NotUnicode(os_string) => {
                    write!(
                        f,
                        "PYTEST_CURRENT_TEST is not a valid unicode string: {os_string:#?}"
                    )
                }
            },
            PytestInfoError::NoTestFile => write!(f, "No test file found"),
        }
    }
}

impl From<PytestInfoError> for PyErr {
    fn from(value: PytestInfoError) -> Self {
        match value {
            PytestInfoError::CouldNotSplit(s) => PyValueError::new_err(format!(
                "Expected '::' to be in PYTEST_CURRENT_TEST string ({s})"
            )),
            PytestInfoError::InvalidEnvVar(ve) => match ve {
                VarError::NotPresent => PyValueError::new_err("PYTEST_CURRENT_TEST is not set"),
                VarError::NotUnicode(os_string) => PyValueError::new_err(format!(
                    "PYTEST_CURRENT_TEST is not a valid unicode string: {os_string:#?}"
                )),
            },
            PytestInfoError::NoTestFile => PyValueError::new_err("No test file found"),
        }
    }
}

#[derive(Debug)]
pub enum SnapError {
    Msg(String),
    Io(std::io::Error),
    Json(serde_json::Error),
    Py(PyErr),
    PytestInfo(PytestInfoError),
    Pythonize(PythonizeError),
    Other(Box<dyn std::error::Error>),
}

impl Display for SnapError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            SnapError::Msg(m) => write!(f, "{m}"),
            SnapError::Io(e) => write!(f, "{e}"),
            SnapError::Json(e) => write!(f, "{e}"),
            SnapError::Py(e) => write!(f, "{e}"),
            SnapError::PytestInfo(e) => write!(f, "{e}"),
            SnapError::Pythonize(e) => write!(f, "{e}"),
            SnapError::Other(e) => write!(f, "{e}"),
        }
    }
}

impl std::error::Error for SnapError {}

impl From<String> for SnapError {
    fn from(value: String) -> Self {
        SnapError::Msg(value)
    }
}

impl From<&str> for SnapError {
    fn from(value: &str) -> Self {
        SnapError::Msg(value.to_string())
    }
}

impl From<std::io::Error> for SnapError {
    fn from(value: std::io::Error) -> Self {
        SnapError::Io(value)
    }
}

impl From<PythonizeError> for SnapError {
    fn from(value: PythonizeError) -> Self {
        SnapError::Pythonize(value)
    }
}

impl From<Box<dyn std::error::Error>> for SnapError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        SnapError::Other(err)
    }
}

impl From<serde_json::Error> for SnapError {
    fn from(value: serde_json::Error) -> Self {
        SnapError::Json(value)
    }
}

impl From<PyErr> for SnapError {
    fn from(value: PyErr) -> Self {
        SnapError::Py(value)
    }
}

impl From<PytestInfoError> for SnapError {
    fn from(value: PytestInfoError) -> Self {
        SnapError::PytestInfo(value)
    }
}

pub type SnapResult<T> = Result<T, SnapError>;
