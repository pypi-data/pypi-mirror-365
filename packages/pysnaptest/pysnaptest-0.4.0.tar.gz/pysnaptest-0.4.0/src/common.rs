use std::collections::BTreeMap;
use std::ops::Deref;
use std::path::PathBuf;
use std::str::{self, FromStr};
use std::sync::{Mutex, MutexGuard};
use std::{env, path::Path};

use once_cell::sync::Lazy;

use pyo3::FromPyObject;
use pyo3::{exceptions::PyValueError, pyclass, pymethods, Bound, PyAny, PyErr, PyResult};

use crate::errors::PytestInfoError;

use insta::internals::{Redaction, SnapshotContents};
use insta::{rounded_redaction, sorted_redaction, Snapshot};
use pyo3::types::PyAnyMethods;

const PYSNAPSHOT_SUFFIX: &str = "pysnap";

static TEST_NAME_COUNTERS: Lazy<Mutex<BTreeMap<String, usize>>> =
    Lazy::new(|| Mutex::new(BTreeMap::new()));

#[derive(Debug)]
struct Description {
    test_file_path: String,
}

impl Description {
    pub fn new(test_file_path: String) -> Self {
        Self { test_file_path }
    }
}

impl From<Description> for String {
    fn from(val: Description) -> Self {
        format!("Test File Path: {}", val.test_file_path)
    }
}

#[derive(Debug)]
pub(crate) struct PytestInfo {
    test_path: String,
    test_name: String,
}

impl PytestInfo {
    pub fn from_env() -> Result<Self, PytestInfoError> {
        let pytest_str = env::var("PYTEST_CURRENT_TEST").map_err(PytestInfoError::InvalidEnvVar)?;
        pytest_str.parse()
    }

    pub fn test_path(&self) -> Result<PathBuf, PytestInfoError> {
        let path = self.test_path_raw();
        if path.exists() {
            Ok(path)
        } else if let Some(filename) = path.file_name() {
            let mut filepath = PathBuf::from("./");
            filepath.push(filename);
            Ok(filepath)
        } else {
            Err(PytestInfoError::NoTestFile)
        }
    }

    pub fn test_path_raw(&self) -> PathBuf {
        Path::new(&self.test_path).to_path_buf()
    }
}

impl FromStr for PytestInfo {
    type Err = PytestInfoError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (test_path, test_name_and_stage) = s
            .split_once("::")
            .ok_or(PytestInfoError::CouldNotSplit(s.to_string()))?;

        let test_name = test_name_and_stage
            .split_once(" ")
            .map(|(tn, _stage)| tn)
            .unwrap_or(test_name_and_stage);

        Ok(PytestInfo {
            test_name: test_name.to_string(),
            test_path: test_path.to_string(),
        })
    }
}

#[pyclass(frozen)]
#[derive(Debug, Clone)]
pub struct SnapshotInfo {
    pub(crate) snapshot_folder: PathBuf,
    pub(crate) snapshot_name: String,
    pub(crate) relative_test_file_path: Option<String>,
    pub(crate) allow_duplicates: bool,
}

impl TryFrom<PytestInfo> for SnapshotInfo {
    type Error = PyErr;
    fn try_from(value: PytestInfo) -> Result<Self, Self::Error> {
        let test_file_dir = value
            .test_path()?
            .canonicalize()?
            .parent()
            .ok_or_else(|| {
                PyValueError::new_err(format!(
                    "Invalid test_path: {:?}, not yielding a parent directory",
                    value.test_path_raw()
                ))
            })?
            .join("snapshots");

        let test_name = &value.test_name;
        let test_path = value.test_path_raw();
        let file_name = test_path.file_stem().and_then(|s| s.to_str());

        let name = if let Some(f) = file_name {
            format!("{f}_{test_name}")
        } else {
            test_name.to_string()
        };
        Ok(Self {
            snapshot_folder: test_file_dir,
            snapshot_name: name,
            relative_test_file_path: Some(value.test_path()?.to_string_lossy().to_string()),
            allow_duplicates: false,
        })
    }
}

impl SnapshotInfo {
    pub(crate) fn counters<'a>() -> MutexGuard<'a, BTreeMap<String, usize>> {
        TEST_NAME_COUNTERS.lock().unwrap_or_else(|x| x.into_inner())
    }

    pub(crate) fn snapshot_name_with_idx(&self, test_idx: usize) -> String {
        if test_idx == 1 || test_idx == 0 {
            self.snapshot_name.to_string()
        } else {
            format!("{}-{}", self.snapshot_name, test_idx)
        }
    }

    pub(crate) fn snapshot_name(&self) -> String {
        let mut c = Self::counters();
        let mut test_idx = c.get(&self.snapshot_name).cloned().unwrap_or(0);
        if !self.allow_duplicates {
            test_idx += 1;
            c.insert(self.snapshot_name.clone(), test_idx);
        }

        self.snapshot_name_with_idx(test_idx)
    }
}

impl TryInto<insta::Settings> for &SnapshotInfo {
    type Error = PyErr;

    fn try_into(self) -> PyResult<insta::Settings> {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path(self.snapshot_folder());
        settings.set_snapshot_suffix(PYSNAPSHOT_SUFFIX);
        if let Some(relative_test_file_path) = &self.relative_test_file_path {
            settings.set_description(Description::new(relative_test_file_path.clone()));
        }
        settings.set_omit_expression(true);
        Ok(settings)
    }
}

#[derive(Debug, Clone)]
pub enum RedactionType {
    Sorted,
    Rounded(usize),
    Standard(String),
}

impl<'source> FromPyObject<'source> for RedactionType {
    #[inline]
    fn extract_bound(ob: &Bound<'source, PyAny>) -> PyResult<Self> {
        if ob.is_none() {
            Ok(RedactionType::Sorted)
        } else if let Ok(decimals) = ob.extract::<usize>() {
            Ok(RedactionType::Rounded(decimals))
        } else if let Ok(redaction) = ob.extract::<String>() {
            Ok(RedactionType::Standard(redaction))
        } else {
            Err(PyErr::new::<pyo3::exceptions::PyTypeError, _>(
                "Unable to extract RedactionType",
            ))
        }
    }
}

impl From<RedactionType> for Redaction {
    fn from(value: RedactionType) -> Self {
        match value {
            RedactionType::Sorted => sorted_redaction(),
            RedactionType::Rounded(decimals) => rounded_redaction(decimals),
            RedactionType::Standard(redaction) => redaction.into(),
        }
    }
}

#[pyclass(unsendable)]
#[derive(Debug)]
pub struct PySnapshot(Snapshot);

#[pymethods]
impl PySnapshot {
    #[staticmethod]
    pub fn from_file(p: PathBuf) -> PyResult<Self> {
        Ok(Self(Snapshot::from_file(&p).map_err(|e| {
            pyo3::exceptions::PyValueError::new_err(format!(
                "Unable to load snapshot from {p:?}, details: {e}",
            ))
        })?))
    }

    pub fn contents(&self) -> PyResult<Vec<u8>> {
        Ok(match self.0.contents() {
            SnapshotContents::Text(text_snapshot_contents) => {
                text_snapshot_contents.to_string().as_bytes().to_vec()
            }
            SnapshotContents::Binary(items) => items.deref().to_owned(),
        })
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_into_pyinfo_happy_path() {
        let s = "tests/a/b/test_thing.py::test_a (call)";
        let pti: Result<PytestInfo, PytestInfoError> = s.parse();
        insta::assert_debug_snapshot!(pti)
    }

    #[test]
    fn test_into_pyinfo_no_trailer() {
        let s = "tests/a/b/test_thing.py::test_a";
        let pti: Result<PytestInfo, PytestInfoError> = s.parse();
        insta::assert_debug_snapshot!(pti)
    }

    #[test]
    fn test_into_pyinfo_failure_case() {
        let s = "tests/a/b/test_thing.py";
        let pti: Result<PytestInfo, PytestInfoError> = s.parse();
        insta::assert_debug_snapshot!(pti)
    }

    #[test]
    fn test_snapshot_info_overrides_from_pytest() {
        let snapshot_info = SnapshotInfo::from_pytest(
            Some("folder_path_override".into()),
            Some("snapshot_name_override".into()),
            false,
        )
        .unwrap();
        insta::assert_debug_snapshot!(snapshot_info);
        insta::assert_snapshot!(snapshot_info.snapshot_name(), @"snapshot_name_override");
        insta::assert_snapshot!(snapshot_info.last_snapshot_name(), @"snapshot_name_override");
        insta::assert_snapshot!(snapshot_info.next_snapshot_name(), @"snapshot_name_override-2");
        insta::assert_snapshot!(snapshot_info.snapshot_name(), @"snapshot_name_override-2");
        insta::assert_snapshot!(snapshot_info.last_snapshot_name(), @"snapshot_name_override-2");
        insta::assert_snapshot!(snapshot_info.next_snapshot_name(), @"snapshot_name_override-3");
    }
}
