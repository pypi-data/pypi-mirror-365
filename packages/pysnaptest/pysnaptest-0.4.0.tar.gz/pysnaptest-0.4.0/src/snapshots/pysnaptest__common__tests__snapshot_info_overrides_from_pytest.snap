---
source: src/common.rs
expression: snapshot_info
---
SnapshotInfo {
    snapshot_folder: "folder_path_override",
    snapshot_name: "snapshot_name_override",
    relative_test_file_path: None,
    allow_duplicates: false,
}
